pkg/auth/dependency/interaction/flows/webapp.go

```diff
@@ -114,6 +114,52 @@ func (f *WebAppFlow) SignupWithLoginID(loginIDKey, loginID string) (*WebAppResul
 	}, nil
 }
 
+func (f *WebAppFlow) PromoteWithLoginID(loginIDKey, loginID string, userID string) (*WebAppResult, error) {
+	i, err := f.Interactions.NewInteractionAddIdentity(&interaction.IntentAddIdentity{
+		Identity: identity.Spec{
+			Type: authn.IdentityTypeLoginID,
+			Claims: map[string]interface{}{
+				identity.IdentityClaimLoginIDKey:   loginIDKey,
+				identity.IdentityClaimLoginIDValue: loginID,
+			},
+		},
+	}, "", userID)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := f.Interactions.GetInteractionState(i)
+	if err != nil {
+		return nil, err
+	} else if len(s.Steps) != 1 || s.Steps[0].Step != interaction.StepSetupPrimaryAuthenticator {
+		panic("interaction_flow_webapp: unexpected interaction state")
+	}
+
+	step := WebAppStepSetupPassword
+	// Trigger OOB OTP if the first authenticator is OOB OTP
+	if len(s.Steps[0].AvailableAuthenticators) > 0 && s.Steps[0].AvailableAuthenticators[0].Type == authn.AuthenticatorTypeOOB {
+		step = WebAppStepSetupOOBOTP
+		err = f.Interactions.PerformAction(i, interaction.StepSetupPrimaryAuthenticator, &interaction.ActionTriggerOOBAuthenticator{
+			Authenticator: s.Steps[0].AvailableAuthenticators[0],
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Line 134 - Line 148 can adjusted according to https://github.com/SkygearIO/skygear-server/pull/1407/commits/ac9091249d2b38c63173156a91a92c008df99add and https://github.com/SkygearIO/skygear-server/pull/1407/commits/0362d54743d12dd7031712c3b150b46ca427901a

Reply:
Ping